### DISCUSSION TITLE
Fixing knownStates

### DISCUSSION THREAD

**Discussion by willpkay** [bmcclintock/momentuHMM#68] [Q&A] [answered]
Hi Brett,

I wish to include knownStates in my HMM. My model has three states: transiting, foraging and resting. I have determined all of the time periods when my animals are assumed to be "resting" (- the data are from seals and following McClintock et al., 2013 and Russell et al., 2015 I have classified proportion of time spent diving using <0.444 threshold).

I hence wish to fit a 3-state model where all of the time intervals representing "resting" states are "known", but allow the model to determine the remaining time intervals as either travelling and foraging states.

I have tried fitting the model using the `knownStates` argument, but when I do this the model not only includes the time intervals that I have specified as known states, but it also fits many more time intervals as additional resting states.

Is there a way to force the model to only assign either travelling or foraging to the remaining time intervals that are not specified in `knownStates`?

This analysis is essentially the same as what you did in Carter et al., 2020 _Oikos_: "_We fitted a three state multivariate, discrete-time HMM with resting as a ‘known state’, and attributed the remaining time intervals with low step lengths and high turn angles to foraging, and intervals with high step lengths and low turn angles to travelling._"

The code I am working with is provided below.

Thank you very much,

Will

<pre>
nbStates<-3 # 3 States (1 = "Transiting", 2 = "Foraging", 3 = "Resting")

dist<-list(step="gamma", 
             angle="wrpcauchy") 

estAngleMean<-list(angle=TRUE)

circularAngleMean<-list(angle=TRUE) 

DM<-list(step= list(mean=~HW*sex*site,
                               sd=~HW), 
                angle=list(mean=~1, 
                                 concentration=~HW*sex*site)) 

formula<-~HW*sex*site

knownStates<-mHMM$knownStates

checkPar0(mHMM,nbStates=nbStates,
          dist=dist,
          DM=DM,
          mixtures=3,
          knownStates=knownStates,
          estAngleMean=estAngleMean,
          larAngleMean=circularAngleMean)

getPar<-list(step=c(0.89,0.64,0.30,0.13,0.15,0.19),
             angle=c(0,0,0,0.6,0.1,0.05))

(getPar<-getParDM(data=mHMM,
                  nbStates=nbStates,
                  dist=dist,
                  Par=getPar,
                  estAngleMean=estAngleMean,
                  knownStates=knownStates,
                  circularAngleMean=circularAngleMean,
                  DM=DM))

Par0<-list(step=getPar$step,
           angle=getPar$angle)

stateNames<-c("Transiting","Foraging","Resting")

m25k<-fitHMM(data=mHMM,
             nbStates=nbStates,
             dist,
             Par0=Par0,
             DM=DM,
             formula=formula,
             mixtures=3,
             knownStates=knownStates,
             estAngleMean=estAngleMean,
             circularAngleMean=circularAngleMean,
             stateNames=stateNames)
</pre>

**Comment by bmcclintock** [comment, marked as answer]
This can be relatively easily done by using the dive threshold indicator as a factor covariate on the initial distribution ('formulaDelta') and state transition probabilities ('formula').  Then by fixing the appropriate parameters to large or small values using fixPar, this would create 2 "sets" of transition probabilities depending on the covariate value (one set that forces transitions to resting when dive % < threshold at next time step and another set that prohibits transition to resting when dive % >= threshold at the next time step):

```
#### Simulate some data
nbObs <- 1000
nbStates <- 3
stateNames <- c("resting","foraging","transit")
dist <- list(step="gamma",angle="vm",dive="beta")

Par <- list(step=c(1,1,10,0.5,0.5,2),
            angle=c(0,0,0,0.1,0.1,10),
            dive=c(1,10,10,10,1,1))

beta <- matrix(-1.5,1,nbStates*(nbStates-1))

simDat <- simData(obsPerAnimal = nbObs,nbStates=nbStates,dist=dist,Par=Par,beta=beta,stateNames=stateNames,states=TRUE)

fitDat <- simDat[-nrow(simDat),] # can't have missing covariates
threshold <- 0.44
fitDat$rest <- factor(as.numeric(fitDat$dive<threshold)) # covariate indicating when resting threshold is satisfied
fitDat$dive <- NULL

#### Set up model
betaRef <- c(2,2,3) # make state 2 (foraging) the tpm reference for state 1 (resting)
beta0 <- matrix(c(-100, 0.5, -100, -1.5, -100, -1.5,
                   100,   0,  100,    0,  100,    0),2,nbStates*(nbStates-1),byrow=TRUE,dimnames=list(c("rest0","rest1"),c("1 -> 1", "1 -> 3", "2 -> 1", "2 -> 3", "3 -> 1", "3 -> 2")))

delta0 <- matrix(c( 100,  100,
                   -100, -100),2,nbStates-1,byrow=TRUE,dimnames=list(c("rest0","rest1"),stateNames[-1]))

formula <- ~ 0 + rest
formulaDelta <- ~ 0 + rest

#### check tpm
trProbs <- getTrProbs(fitDat[,c("ID","rest"),drop=FALSE],nbStates=nbStates,beta=beta0,formula=formula,betaRef=betaRef,stateNames=stateNames)
trProbs[,,which.min(fitDat$rest)] # tpm when next step has rest=0
trProbs[,,which.max(fitDat$rest)] # tpm when next step has rest=1

Par0 <- Par[c("step","angle")]
Par0$angle <- Par$angle[-(1:nbStates)]

#### Set knownStates and fixPar$beta for extra insurance (for even extra insurance could also set workBounds$beta to be [-Inf,<<100] or [>>-100,Inf] as appropriate for estimated tpm parameter)
knownStates <- rep(NA,nbObs-1)
knownStates[fitDat$rest==1] <- 1
fixPar <- list(beta=beta0)
fixPar$beta[which(beta0 > -100 & beta0 < 100 & beta0!=0)] <- NA
workBounds <- list(delta=matrix(c(100-1.e-6,-Inf,100-1.e-6,-Inf,Inf,-100+1.e-6,Inf,-100+1.e-6),nrow=4,ncol=2))
fixPar$delta <- momentuHMM:::nw2w(delta0,workBounds$delta)
if(fitDat$rest[1]==0){
  fixPar$delta[1,] <- NA
}

#### fit the model
fit <- fitHMM(fitDat,nbStates=nbStates,dist=dist[c("step","angle")],Par0=Par0,beta0=beta0,formula=formula,formulaDelta=formulaDelta,betaRef=betaRef,fixPar=fixPar,workBounds=workBounds,knownStates=knownStates,stateNames=stateNames)
fit
plot(fit,plotCI=TRUE,ask=FALSE)
unique(viterbi(fit)[which(fitDat$rest==0)])
unique(viterbi(fit)[which(fitDat$rest==1)])
```

**Comment by willpkay** [comment]
Great, I will give this a try. Thanks a lot Brett, especially for the super quick reply!

**Comment by willpkay** [comment]
Hi again Brett,

Apologies for following up with another question. I understand the theory which seems elegant/intuitive and relatively straight-forward, however I am struggling with the implementation in R. 

Specifically, I am unsure how to specify `beta0` and `delta0` correctly - In my model I wish to include three covariates in addition to `rest`: 

- `HW` (continuous; -0.9 to 1.15)
- `sex` (factorial; F or M), and 
- `site` (factorial; DK, LP, or HE). 

I would like to model their effects as full interactions on the state transition probabilities i.e. `formula = ~ HW * sex * site`. 

However, I am now not sure how to correctly include these in `beta0` and `delta0` in addition to the newly introduced `rest` covariate. Would it be as simple as: `formula <- ~ HW * sex * site + rest`

My first attempt was as follows:

<pre>
#### Set up model
nbStates <- 3

stateNames <- c("resting","foraging","transit")

dist <- list(step="gamma",angle="vm",dive="beta")

Par <- list(step=c(0.3, 0.64, 0.89, 0.19, 0.15, 0.13),
            angle=c(0, 0, 0, 0.05, 0.1, 0.6),
            dive=c(1, 10, 10, 10, 1, 1))

beta <- matrix(-1.5, 1, nbStates*(nbStates-1))

betaRef <- c(2,2,3) # make state 2 (foraging) the tpm reference for state 1 (resting)

beta0 <- matrix(c(-100, 0.5, -100, -1.5, -100, -1.5, 
                   100,   0,  100,    0,  100,    0,
                     0,   0,    0,    0,    0,    0,
                     0,   0,    0,    0,    0,    0,
                     0,   0,    0,    0,    0,    0), 5, nbStates*(nbStates-1), 
                byrow=TRUE,
                dimnames=list(c("rest0","rest1","HW","sex","site"),
                              c("1 -> 1", "1 -> 3", "2 -> 1", "2 -> 3", "3 -> 1", "3 -> 2")))

delta0 <- matrix(c(100,  100,
                  -100, -100,
                     1,    1,
                     1,    1,
                     1,    1), 5, nbStates-1,
                 byrow=TRUE,
                 dimnames=list(c("rest0","rest1","HW","sex","site"), stateNames[-1]))

formula <- ~ HW * sex * site + rest
formulaDelta <- ~ HW * sex * site + rest

#### check tpm
trProbs <- getTrProbs(mHMM[,c("rest","HW","sex","site"), drop=FALSE],
                      nbStates=nbStates,
                      beta=beta0,
                      formula=formula,
                      betaRef=betaRef,
                      stateNames=stateNames)
</pre>

This however returns an error stating: 
<pre>
Error in getTrProbs.default(mHMM[, c("rest", "HW", "sex", "site"), drop = FALSE],  : 
  beta must be a matrix with13rows and6columns
</pre>

Hence, I amended beta0 and delta0 to include 13 rows, though I am not sure if I have specified the covariates correctly(?). I anticipate that I need to include all covariates and their interactions (i.e. `HW`, `sex`, `site`, `HW:sex`, `HW:site`, `sex:site`, `HW:sex:site`), plus the intercepts of `rest0` and `rest1`. This however adds up to only 9 rows. Am I correct in thinking the remaining 4 rows should be attributed to the interaction between rest and the other covariates? i.e. `rest:HW`, `rest:sex`, `rest:site`, `rest:HW:sex:site`, Or perhaps I have completely misunderstood..

I attempted the following:

<pre>
beta0 <- matrix(c(-100, 0.5, -100, -1.5, -100, -1.5, 
                   100,   0,  100,    0,  100,    0,
                     0,   0,    0,    0,    0,    0,
                     0,   0,    0,    0,    0,    0,
                     0,   0,    0,    0,    0,    0,
                     0,   0,    0,    0,    0,    0,
                     0,   0,    0,    0,    0,    0,
                     0,   0,    0,    0,    0,    0,
                     0,   0,    0,    0,    0,    0,
                     0,   0,    0,    0,    0,    0,
                     0,   0,    0,    0,    0,    0,
                     0,   0,    0,    0,    0,    0,
                     0,   0,    0,    0,    0,    0), 13, nbStates*(nbStates-1), 
                byrow=TRUE,
                dimnames=list(c("rest0","rest1",
                                "HW","sex","site",
                                "HW:sex","HW:site","sex:site","HW:sex:site",
                                "rest:HW","rest:sex","rest:site","rest:HW:sex:site"),
                              c("1 -> 1", "1 -> 3", "2 -> 1", "2 -> 3", "3 -> 1", "3 -> 2")))

delta0 <- matrix(c(100,  100,
                  -100, -100,
                     0,    0,
                     0,    0,
                     0,    0,
                     0,    0,
                     0,    0,
                     0,    0,
                     0,    0,
                     0,    0,
                     0,    0,
                     0,    0,
                     0,    0), 13, nbStates-1,
                 byrow=TRUE,
                 dimnames=list(c("rest0","rest1",
                                 "HW","sex","site",
                                 "HW:sex","HW:site","sex:site","HW:sex:site",
                                 "rest:HW","rest:sex","rest:site","rest:HW:sex:site"), stateNames[-1]))
</pre>

This allowed the `getTrProbs` function to run, however the resulting tpm for when rest=1 is incorrect:

<pre>
trProbs <- getTrProbs(mHMM[,c("rest","HW","sex","site"), drop=FALSE],
                      nbStates=nbStates,
                      beta=beta0,
                      formula=formula,
                      betaRef=betaRef,
                      stateNames=stateNames)

trProbs[,,which.min(mHMM$rest)] # tpm when next step has rest=0
              resting  foraging   transit
resting  2.870985e-31 0.3775407 0.6224593
foraging 6.217195e-31 0.8175745 0.1824255
transit  6.217195e-31 0.1824255 0.8175745

trProbs[,,which.max(mHMM$rest)] # tpm when next step has rest=1
              resting  foraging   transit
resting  1.090426e-43 0.3775407 0.6224593
foraging 2.361347e-43 0.8175745 0.1824255
transit  2.361347e-43 0.1824255 0.8175745
</pre>

I have continued with the rest of the code to try to get it working and determine what am I doing wrong, but likewise I am struggling to understand how I should be specifying `fixPar` and `workBounds` with my covariates.

I have looked through the momentuHMM vignette and see that the African buffalo recharge dynamics example includes user-specification of reference states, but this model seems quite different to what I am trying to achieve here. Is there an example in particular that you would recommend I cross-reference?

Thanks again,

Will

**Comment by bmcclintock** [comment]
Wow that's a lot of interactions! Might want to be careful there unless you've got a really great data set. I suggest you start with a single covariate (`sex`, say) and play around with `checkPar0` and `getTrProbs` until you get it right. Once you figure out how to do it with one covariate, I suspect you'll understand what's going on and will then be able to add additional covariates.

**Comment by willpkay** [comment]
Indeed! OK, will do. Thanks again!
